index.js

```diff
@@ -243,12 +275,52 @@ Shopify.prototype.graphql = function graphql(data, variables) {
     parseJson: this.options.parseJson,
     timeout: this.options.timeout,
     responseType: 'json',
-    retry: 0,
     method: 'POST',
-    body: json ? this.options.stringifyJson({ query: data, variables }) : data
+    body: json ? this.options.stringifyJson({ query: data, variables }) : data,
+    retry:
+      this.options.maxRetries > 0
+        ? {
+            limit: this.options.maxRetries,
+            // Don't clamp Shopify retry-after header values too low.
+            maxRetryAfter: Infinity,
+            calculateDelay: (retryObject) => {
+              attemptNumber += 1;
+              return (
+                graphQLRetryMS(
+                  attemptNumber,
+                  this.options.maxRetries,
+                  retryObject.error
+                ) || retryObject.computedValue
+              );
+            }
+          }
+        : 0
   };
 
-  return got(uri, options).then((res) => {
+  const instance = got.extend({
+    hooks: {
+      afterResponse: [
+        (response, retryWithMergedOptions) => {
```

Comment:
I think this hook should not be added when the `autoLimit` option is set. Also, instead of using `got.extend()` isn't it possible to add the `hooks` object in the `options` object?

Reply:
I tried as well to get rid of the `got.extend`, but you need a reference to the instance to grab the options to pass to the `retryWithMergedOptions` function. It pukes if you don't send in the same instance's options or no options at all. This `instance =` pattern is copied from the example here: https://github.com/sindresorhus/got/blob/main/documentation/9-hooks.md#afterresponse

Reply:
Hmm even if you pass an empty options object?

Reply:
Only one test fails with this:

```diff
$ git diff
diff --git a/index.js b/index.js
index 82a4a0f..5bd8873 100644
--- a/index.js
+++ b/index.js
@@ -297,33 +297,29 @@ Shopify.prototype.graphql = function graphql(data, variables) {
         : 0
   };
 
-  const instance = got.extend({
-    hooks: {
-      afterResponse:
-        this.options.maxRetries > 0
-          ? [
-              (response, retryWithMergedOptions) => {
-                if (response.body && response.body.errors) {
-                  const retryDelay = graphQLRetryMS(
-                    attemptNumber,
-                    this.options.maxRetries,
-                    { response }
-                  );
-
-                  if (retryDelay) {
-                    return retryWithMergedOptions(instance.default.options);
-                  }
-                }
-
-                return response;
-              }
-            ]
-          : []
-    },
-    mutableDefaults: true
-  });
+  if (this.options.maxRetries > 0) {
+    options.hooks = {
+      afterResponse: [
+        (response, retryWithMergedOptions) => {
+          if (response.body && response.body.errors) {
+            const retryDelay = graphQLRetryMS(
+              attemptNumber,
+              this.options.maxRetries,
+              { response }
+            );
+
+            if (retryDelay) {
+              return retryWithMergedOptions(options);
+            }
+          }
+
+          return response;
+        }
+      ]
+    };
+  }
 
-  return instance(uri, options).then((res) => {
+  return got(uri, options).then((res) => {
     if (res.body.extensions && res.body.extensions.cost) {
       this.updateGraphqlLimits(res.body.extensions.cost);
     }
```

Reply:
Nice catch -- definitely not on purpose. Instead of trying to figure out what strange options we need to pass to that retry function, I updated this to throw, which I actually think matches what is happening better. We've got a response but we really did encounter an error condition (if this mode is configured), so throwing seems like a nice and explicit way of triggering a retry. I tried to get that retry function working but it requires the ugly `instance =` stuff and an options merge that seems more complicated than it is worth.